### ISSUE TITLE
Deterministic operations

### ISSUE THREAD

**Issue by yorard123** [rusty1s/pytorch_geometric#446]
Hi,

I'm wondering if it is possible to implement in a deterministic way all scatter operation even if it means sacrificing speed. I'm asking that because I found that using the same random seed there is a lot of difference between the results obtaining in different runs (in some of my cases I had 4% of accuracy difference 69% to 73%). For this reason, sometimes it is difficult to verify if the changes have a positive impact on the results or it is only luck. 

I'm not sure if I completely explained the problem that I am facing. 

Thanks,

**Comment by rusty1s** [comment]
Hi, a deterministic scatter operation (on GPU) does not exist as far as I know. This is only possible if the input is coalesced (and you implement `scatter_*` as `segment_*`). This is sadly currently not possible with the means of PyG (and I never witnessed such high jumps in accuracy by myself).

**Comment by yorard123** [comment]
Hi,
I'm not sure if [this](https://huangjianyu.wordpress.com/2015/03/14/parallel-pattern-10-scatter/) can be applied to PyG (I am not versed in gpu programming) but maybe it is useful. 

> and I never witnessed such high jumps in accuracy by myself

I think it is only happening in complex datasets. I will try to reproduce the issue in a simple dataset(I can not upload the dataset I am using is too huge.) In the midtime, I would like to show you the results that I am obtaining. Using the same random seed in both experiments.
Train Metrics:
pink=exp1
orange=exp2
**Acc**
![train_acc](https://user-images.githubusercontent.com/48563127/60155927-9dad4d80-97eb-11e9-8211-0b1f6dfe4563.png)
**Loss**
![train_loss](https://user-images.githubusercontent.com/48563127/60155934-a1d96b00-97eb-11e9-9791-c6d64cc0fc4e.png)

As you can see both curves have more or less the same behavior in training but the issue happens in the validation set:
Blue= exp1 75.18%
green= exp2 73.68%
**Acc**
![val_acc](https://user-images.githubusercontent.com/48563127/60156041-e8c76080-97eb-11e9-8e79-4dc4a712a575.png)
**Loss**
![loss_val](https://user-images.githubusercontent.com/48563127/60156049-ebc25100-97eb-11e9-9e82-3cd37723c943.png)

As you can see the behavior in the validation set is very different even though using the same random seed.

**Comment by rusty1s** [comment]
Do you obtain same accuracies when training on CPU only? What happens if you change from `torch.float` to `torch.double`? Does this fix some of these issues?

**Comment by yorard123** [comment]
> Do you obtain same accuracies when training on CPU only? 
I didn't try, I'll do it

> What happens if you change from torch.float to torch.double? Does this fix some of these issues?
I can not do it. I don't have enough GPU memory. I will try to make a more simple model and try this.

**Comment by yorard123** [comment]
> Do you obtain same accuracies when training on CPU only?

At this moment I ran 4 different experiments with the same seed and hyperparametres and the first epoch train/val has the same exact accuracy. I will run the experiments for 10 epochs to be sure that the results are exactly the same. However, each epoch needs 1 hour to complete... 

> What happens if you change from torch.float to torch.double? Does this fix some of these issues?

I managed to fit this change to the gpu and I can confirm that change the float to double is not solving the issue.

**Comment by rusty1s** [comment]
I thought that at least double precision would prevent this behaviour to a certain extent. Well, due to concurrent thread this problem is in the nature of GPU processing, and, e.g., TensorFlow has similar problems. In the end, there is not much I can do about it, although I wonder how CUDNN implements GPU determinism.

**Comment by amosella** [comment]
Hi,
I'm sorry for interrupting your conversation. I was facing a similar behavior days ago. Like you, I wondered how some implementations of cudnn has not have these kinds of issues and I found this [article](https://people.eecs.berkeley.edu/~hdnguyen/public/papers/ARITH21_Fast_Sum.pdf). The basic idea is that they try to cancel out the lowest bits(that are indeterministic). I'm not sure if this is the pattern used it in CUDNN but maybe it can be a solution for this problem. What do you think?

**Comment by rusty1s** [comment]
Yes, we just use the `scatter_add` operation from PyTorch, and implement other scatter functions analogously.

**Comment by amosella** [comment]
I think it can be possible to implement the methodology described in the previous article using the existing operations.

**Comment by rusty1s** [comment]
It would be nonetheless great to have the PyTorch team on board with this one, so creating an issue there might be a good option.

**Comment by amosella** [comment]
I already did it, [here](https://discuss.pytorch.org/t/possible-solution-to-the-reproducibility-issues-using-scatter-add-operation/48989)

**Comment by yorard123** [comment]
After three epochs all runs are giving the same outputs (in cpu). I suppose I can confirm that the issue is cpu related. I doubled checked the use of doubles, and it is not solving the issue :( I suppose we can only wait for the pytorch team answer. Thanks.

**Comment by amosella** [comment]
There is some [advance](https://discuss.pytorch.org/t/possible-solution-to-the-reproducibility-issues-using-scatter-add-operation/48989) in the discussion, but it seems if we want to have deterministic behaviour we will need to do it ourselves.

**Comment by rusty1s** [comment]
Thank you for this insightful discussion, I will look at the kernels in more detail at some later time. It seems to me that having deterministic `scatter_add` functionality will come with a huge runtime trade-off. I think your linked paper might be a better approach to solving this task  (also for runtime) than using something like sorting, as we would basically implement [`segment_sum`](https://www.tensorflow.org/api_docs/python/tf/math/segment_sum) this way (a functionality which is sadly missed in PyTorch also).

**Comment by amosella** [comment]
> Thank you for this insightful discussion, I will look at the kernels in more detail at some later time. It seems to me that having deterministic scatter_add functionality will come with a huge runtime trade-off.

I have the same feeling.

> I think your linked paper might be a better approach to solving this task (also for runtime) than using something like sorting, as we would basically implement segment_sum this way (a functionality which is sadly missing in PyTorch also)

I agree with you that in terms of runtime it seems to be better the paper I linked.

This weekend I will try to spend some time on this topic. I think I will start to try to imitate the functionality of the kernels provided. (As I said on the discussion I am not fluent in cuda, I think for me it would be easier to start with this).

Maybe we can keep this issue to track the advances on this topic?

**Comment by rusty1s** [comment]
> Maybe we can keep this issue to track the advances on this topic?

Sure, feel free to write here any time. I am really thankful for your help on this one :)

**Comment by amosella** [comment]
Sorry for the delay. The last weeks were a bit crazy for me :( I hope that in august I will have more free time to continue on this.  Moreover, I started an issue in PyTorch GitHub as was proposed yesterday in the thread I started weeks ago.
